docs/user_guide/config_options.md
# config.yaml options

The different running modes of `benchcab` are solely dependent on the options used in `config.yaml`. The following gives some typical ways to configure `benchcab` for each mode, but the tool is not restricted to these choices of options:

=== "Regression test"

    For this test, you want to:

    * Specify the details of two branches of CABLE
    * Do not specify a [`patch`](#+realisation.patch)
    * Use the default set of science options, i.e. do not specify [`science_configurations`](#`science_configurations`) in `config.yaml`
    * Choose the [`experiment`](#`experiment`) suitable for your stage of development. A run with the `forty-two-site-test` will be required for submissions of new development to CABLE.

=== "New feature test"

    For this test, you want to:

    * Specify the details of two branches of CABLE
    * Specify a [`patch`](#+realisation.patch) for **one** of the branches
    * Use the default set of science options, i.e. do not specify [`science_configurations`](#`science_configurations`) in `config.yaml`
    * Choose the [`experiment`](#`experiment`) suitable for your stage of development. A run with the `forty-two-site-test` will be required for submissions of new development to CABLE.


=== "Ensemble mode"

    This running mode is quite open to customisations:

    * Specify the number of CABLE's branches you need
    * Use [`patch`](#+realisation.patch) on branches as required
    * Specify the [science configurations](#`science_configurations`) you want to run. [`patch`](#+realsation.patch) will be applied on top of the science configurations listed.


## project

: **Default:** user's default project, _optional key_. :octicons-dash-24: NCI project ID to charge the simulations to. The user's default project defined in the $PROJECT environment variable is used by default.

``` yaml

project: nf33

```

## modules

: **Default:** _required key, no default_. :octicons-dash-24: NCI modules to use for compiling CABLE

``` yaml

modules: [
  intel-compiler/2021.1.1,
  netcdf/4.7.4,
  openmpi/4.1.0
]

```

## fluxsite
Contains settings specific to fluxsite tests.

This key is _optional_. **Default** settings for the fluxsite tests will be used if it is not present

```yaml
fluxsite:
  experiment: AU-How
  pbs:
    ncpus: 18
    mem: 30GB
    walltime: 06:00:00
    storage: [scratch/a00, gdata/xy11]
  multiprocess: True
```

### [experiment](#experiment)

: **Default:** `forty-two-site-test`, _optional key_. :octicons-dash-24: Type of fluxsite experiment to run. Experiments are defined in the **benchcab-evaluation** workspace on [modelevaluation.org][meorg]. This key specifies the met forcing files used in the test suite. To choose from:

: | Key value | Experiment description |
|-----------|------------------------|
| [`forty-two-site-test`][forty-two-me] | Run simulations for 42 FLUXNET sites |
| [`five-site-test`][five-me] | Run simulations for 5 FLUXNET sites |
| `AU-Tum` | Run simulations at the Tumbarumba (AU) site |
| `AU-How` | Run simulations at the Howard Spring (AU) site |
| `FI-Hyy` | Run simulations at the Hyytiala (FI) site |
| `US-Var` | Run simulations at the Vaira Ranch-Ione (US) site |
| `US-Whs` | Run simulations at the Walnut Gulch Lucky Hills Shrub (US) site |

```yaml
fluxsite:
  experiment: AU-How

```

### [pbs](#pbs)

Contains settings specific to the PBS scheduler at NCI for the PBS script running the CABLE simulations at FLUXNET sites and the bitwise comparison for these simulations.

This key is _optional_. **Default** values for the PBS settings will apply if it is not specified.

```yaml
fluxsite:
  pbs:
    ncpus: 18
    mem: 30GB
    walltime: 06:00:00
    storage: [scratch/a00, gdata/xy11]
```

[`ncpus`](#+pbs.ncpus){ #+pbs.ncpus }

: **Default:** 18, _optional key_. :octicons-dash-24: The number of CPU cores to allocate for the PBS job, i.e. the `-l ncpus=<4>` PBS flag in [PBS Directives Explained][nci-pbs-directives].

```yaml

fluxsite:
  pbs:
    ncpus: 18

```

[`mem`](#+pbs.mem){ #+pbs.mem }

: **Default:** 30GB, _optional key_. :octicons-dash-24: The total memory limit for the PBS job, i.e. the `-l mem=<10GB>` PBS flag in [PBS Directives Explained][nci-pbs-directives].

```yaml

fluxsite:
  pbs:
    mem: 30GB

```

[`walltime`](#+pbs.walltime){ #+pbs.walltime }

: **Default:** `6:00:00`, _optional key_. :octicons-dash-24: The wall clock time limit for the PBS job, i.e. `-l walltime=<HH:MM:SS>` PBS flag in [PBS Directives Explained][nci-pbs-directives].

```yaml

fluxsite:
  pbs:
    walltime: 6:00:00

```
[`storage`](#+pbs.storage){ #+pbs.storage }

: **Default:** [], _optional key_. :octicons-dash-24: List of extra storage flags required for the PBS job, i.e. `-l storage=<scratch/a00>` in [PBS Directives Explained][nci-pbs-directives].

```yaml

fluxsite:
  pbs:
    storage: [scratch/a00, gdata/xy11]

```

### [multiprocess](#multiprocess)

: **Default:** True, _optional key_. :octicons-dash-24: Enables or disables multiprocessing for executing embarrassingly parallel tasks.

```yaml

fluxsites:
  multiprocess: True

```

## spatial

Contains settings specific to spatial tests.

This key is _optional_. **Default** settings for the spatial tests will be used if it is not present.

```yaml
spatial:
  met_forcings:
    crujra_access: https://github.com/CABLE-LSM/cable_example.git
  payu:
    config:
      walltime: 1:00:00
    args: -n 2
```

### [met_forcings](#met_forcings)

Specify one or more spatial met forcings to use in the spatial test suite. Each entry is a key-value pair where the key is the name of the met forcing and the value is a URL to a payu experiment that is configured to run CABLE with that forcing.

This key is _optional_. **Default** values for the `met_forcings` key is as follows:

```yaml
spatial:
  met_forcings:
    crujra_access: https://github.com/CABLE-LSM/cable_example.git
```

### [payu](#payu)

Contains settings specific to the payu workflow manager.

This key is _optional_. **Default** values for the payu settings will apply if not specified.

```yaml
spatial:
  payu:
    config:
      walltime: 1:00:00
    args: -n 2
```

[`config`](#+payu.config){ #+payu.config }

: **Default:** unset, _optional key_. :octicons-dash-24: Specify global configuration options for running payu. Settings specified here are passed into to the payu configuration file for each experiment.

```yaml
spatial:
  payu:
    config:
      walltime: 1:00:00
```

[`args`](#+payu.args){ #+payu.args }

: **Default:** unset, _optional key_. :octicons-dash-24: Specify command line arguments to the `payu run` command in the form of a string. Arguments are used for all spatial payu runs.

```yaml
spatial:
  payu:
    args: -n 2
```

## realisations

Entries for each CABLE branch to use. Each entry is a key-value pair and are listed as follows:

```yaml
realisations:
  # head of main branch
  - repo:
      git:
        branch: main
  # some development branch
  - repo:
      git:
        branch: my_branch
    patch:
      cable:
        cable_user:
          FWSOIL_SWITCH: "Lai and Ktaul 2000"
    patch_remove:
      cable:
        soilparmnew: nil
```

### [repo](#repo)

Contains settings to specify the CABLE branch to test against. 

This key is _required_. The `repo` key must specify the [`svn`](#+repo.svn), the [`git`](#+repo.git) or the [`local`](#+repo.local) key.

```yaml
realisations:
  - repo:
      svn:
        branch_path: trunk
  - repo:
      git:
        branch: main
  - repo:
      local:
        path: /home/ab1234/cable_local
```

#### [`svn`](#+repo.svn){ #+repo.svn}

Contains settings to specify a branch from the CABLE SVN repository (`https://trac.nci.org.au/svn/cable`).

This key is _optional_. No default.

```yaml
realisations:
  - repo:
      svn:
        branch_path: branches/Users/foo/my_branch
        revision: 1234
```

[`branch_path`](#+repo.svn.branch_path){ #+repo.svn.branch_path}

: **Default:** _required key, no default_. :octicons-dash-24: Specify the branch path relative to the SVN root of the CABLE repository (`https://trac.nci.org.au/svn/cable`).

```yaml
realisations:
  # head of the trunk
  - repo:
      svn:
        branch_path: trunk # (1)
  # some development branch
  - repo:
      svn:
        branch_path: branches/Users/foo/my_branch # (2)
```

1. To checkout `https://trac.nci.org.au/svn/cable/trunk`
2. To checkout `https://trac.nci.org.au/svn/cable/branches/Users/foo/my_branch`

[`revision`](#+repo.svn.revision){ #+repo.svn.revision}

: **Default:** HEAD of the branch is checked out, _optional key_. :octicons-dash-24: Specify the revision number to checkout for the branch. This option can be used to ensure the reproducibility of the tests.

```yaml
realisations:
  - repo:
      svn:
        branch_path: branches/Users/foo/my_branch
        revision: 1234
```

#### [`git`](#+repo.git){ #+repo.git}

Contains settings to specify a branch on the GitHub repository. By default, the [CABLE GitHub repository][cable-github] will be cloned (see [`url`](#+repo.git.url) to specify another GitHub repository).

This key is _optional_. No default.

```yaml
realisations:
  - repo:
      git:
        branch: my_branch
        commit: 067b1f4a570385fce01552fdf96ced0adbbe17eb
        url: https://github.com/SeanBryan51/CABLE.git
```

[`branch`](#+repo.git.branch){ #+repo.git.branch}

: **Default:** _required key, no default_. :octicons-dash-24: Specify the GitHub branch name to be checked out.

```yaml
realisations:
  - repo:
      git:
        branch: my_branch
```

[`commit`](#+repo.git.commit){ #+repo.git.commit}

: **Default:** unset, _optional key_. :octicons-dash-24: Specify a specific commit to use for the branch.

```yaml
realisations:
  - repo:
      git:
        branch: my_branch
        commit: 067b1f4a570385fce01552fdf96ced0adbbe17eb
```

[`url`](#+repo.git.url){ #+repo.git.url}

: **Default:** URL of the [CABLE GitHub repository][cable-github], _optional key_. :octicons-dash-24: Specify the GitHub repository url to clone from when checking out the branch.

#### [`local`](#+repo.local){ #+repo.local}

Contains settings to specify CABLE checkouts on a local repository.

This key is _optional_. No default.

```yaml
realisations:
  - repo:
      local:
        path: /scratch/tm70/ab1234/CABLE
```

[`path`](#+repo.local.path){ #+repo.local.path}

: **Default:** _required key, no default_. :octicons-dash-24: Specify the local checkout path of CABLE branch.

### [meorg_output_name](#meorg_output_name)


: **Default:** False, _optional key_. :octicons-dash-24: If this flag is enabled for a given realisation, benchcab will create a model output on [modelevaluation.org][meorg] which is named using the selected realisation, and upload all simulation results to the created model output. Note that the CABLE branch name of the selected realisation should follow the GiHub issue branch format (i.e. it should start with a digit, with words separated by dashes), and must not exceed 50 characters in length. If the branch name does not satisfy these requirements by default, the branch name can be set to a custom branch name via the [`name` option](#name). More precisely, the following steps are executed when enabling this setting:

1. A model output on [modelevaluation.org][meorg] is created with name `<realisation_name>-<hash>` where `realisation_name` is the branch name of the selected realisation, and `hash` is a 6-character hash unique to the `realisations_name`, `model_profile_id` and `$USER` (used to prevent name conflicts for different users' needs).

    **Note**: In case the model output already exists on [modelevaluation.org][meorg], the files within that model output are deleted. This is done to send a fresh set of benchmarking results for analysis, ensuring that the user can re-run `benchcab` without any issues.

2. The following settings are taken by default for the model output:
    * Model Profile - `CABLE`
    * State Selection - `default`
    * Parameter Selection - `automated`
    * Bundled experiments - `true`
    * Comments - `none`

3. The appropriate experiment and empirical model benchmarks on [modelevaluation.org][meorg] are linked to the created model output.

5. The analysis is triggered remotely on [modelevaluation.org][meorg].

```yaml
realisations:
  - repo:
      git:
        branch: main
  - repo:
      git:
        branch: 123-my-branch
    meorg_output_name: True
```

### [name](#name)

: **Default:** base name of [branch_path](#+repo.svn.branch_path) if an SVN repository is given; the branch name if a git repository is given; the folder name if a local path is given, _optional key_. :octicons-dash-24: An alias name used internally by `benchcab` for the branch. The `name` key also specifies the directory name of the source code when retrieving from SVN, GitHub or local.

```yaml
realisations:
  - repo:
      git:
        branch: my_branch
    name: my_feature # (1)
```

1. Checkout the branch in the directory `src/my_feature`

### [build_script](#build_script)

: **Default:** unset, _optional key_. :octicons-dash-24: The path to a custom shell script to build the code in that branch, relative to the repository root directory. **Note:** any lines in the provided shell script that call the [environment modules API][environment-modules] will be ignored. To specify modules to use for the build script, please specify them using the [`modules`](#modules) key.

```yaml
realisations:
  - repo:
      git:
        branch: my_branch
    build_script: offline/build.sh
```

### [install_dir](#install-dir)

: **Default:** unset, _optional key_. :octicons-dash-24: The path to the directory containing the installed executables relative to the project root directory of the CABLE repository. If specified, `benchcab` will look for executables in this directory when building up the run directories.

```yaml
realisations:
  - repo:
      git:
        branch: my_branch
    install_dir: path/to/bin/directory
```

### [patch](#patch)

: **Default:** unset, _optional key_. :octicons-dash-24: Branch-specific namelist settings for `cable.nml`. Settings specified in `patch` get "patched" to the base namelist settings used for both branches. Any namelist settings specified here will overwrite settings defined in the default namelist file and in the science configurations. This means these settings will be set as stipulated in the `patch` for this branch for all science configurations run by `benchcab`.
: The `patch` key must be a dictionary-like data structure that is compliant with the [`f90nml`][f90nml-github] python package.
: To specify a boolean namelist option, one needs to use the Python booleans, True and False, **not the Fortran booleans**

```yaml
realisations:
  - repo:
      git:
        branch: my_branch
    patch:  # (1)
      cable:
        cable_user:
          FWSOIL_SWITCH: "Lai and Ktaul 2000"
          litter: True # (2)
```

1. Sets FWSOIL_SWITCH to "Lai and Ktaul 2000" for all science configurations for this branch
2. The Python boolean will be translated to a Fortran boolean when writing to the namelist file
   
### [patch_remove](#patch_remove)

: **Default:** unset, _optional key. :octicons-dash-24: Specifies branch-specific namelist settings to be removed from the `cable.nml` namelist settings. When the `patch_remove` key is specified, the specified namelists are removed from all namelist files for this branch for all science configurations run by `benchcab`. When specifying a namelist parameter in `patch_remove`, the value of the namelist parameter is ignored.
: The `patch_remove` key must be a dictionary-like data structure that is compliant with the [`f90nml`][f90nml-github] python package.

```yaml
realisations:
  - repo:
      git:
        branch: my_branch
    patch_remove:
      cable:
        soilparmnew: nil # (1)
```

1. The value is ignored and does not have to be a possible value for the namelist option.


## science_configurations

: **Default:** unset, _optional key_. :octicons-dash-24: User defined science configurations. Science configurations that are specified here will replace [the default science configurations](default_science_configurations.md). In the output filenames, each configuration is identified with S<N\> where N is an integer starting from 0 for the first listed configuration and increasing by 1 for each subsequent configuration.
: To specify a boolean namelist option, one needs to use the Python booleans, True and False, **not the Fortran booleans**

```yaml
science_configurations: [
  { # S0 configuration
    cable: {
      cable_user: {
        GS_SWITCH: "medlyn",
        FWSOIL_SWITCH: "Haverd2013",
        litter: True # (1)
      }
    }
  },
  { # S1 configuration
    cable: {
      cable_user: {
        GS_SWITCH: "leuning",
        FWSOIL_SWITCH: "Haverd2013"
      }
    }
  }
]
```

1. The Python boolean will be translated to a Fortran boolean when writing the namelist file.

## codecov

: **Default:** False, _optional key. :octicons-dash-24: Specifies whether to build `benchcab` with code-coverage flags, which can then be used in post-run analysis (`benchcab gen_codecov`).

```yaml
codecov:
  true
```

[meorg]: https://modelevaluation.org/
[forty-two-me]: https://modelevaluation.org/experiment/display/s6k22L3WajmiS9uGv
[five-me]: https://modelevaluation.org/experiment/display/Nb37QxkAz3FczWDd7
[f90nml-github]: https://github.com/marshallward/f90nml
[environment-modules]: https://modules.sourceforge.net/
[nci-pbs-directives]: https://opus.nci.org.au/display/Help/PBS+Directives+Explained
[cable-github]: https://github.com/CABLE-LSM/CABLE

## meorg_bin

: **Default:** False, _optional key. :octicons-dash-24: Specifies the absolute system path to the ME.org client executable. In the absence of this key it will be inferred from the same directory as benchcab should `meorg_output_name` be set in `realisations` above.

``` yaml

meorg_bin: /path/to/meorg

```


ChangeLog
CHANGES
=======

* new version
* Add specification of a subset of met. sites
* Handling of revision number for branches
* Formatting

v0.1.5
------

* Update version
* Log last change revision number for each branch
* Add rev. number in svn checkout and config.yaml
* Remove code to create user copy of trunk
* new version
* Change source of namelist files Fixes #13
* Need the soil namelist in the run dir
* Changelog

v0.1.4
------

* Update version
* Get the pft\_params.nml file in run directory
* Clarify instructions for run directory

v0.1.3
------

* Update version
* Fixes issue #1: missing line to subprocess

v0.1.2
------

* Update README.md
* Update README.md
* Update README.md
* Update README.md
* Add file auto-created
* Revised job size for better CPU utilisation
* Correct PyYAML dependency for conda
* Update version
* Add yaml dependency for conda
* Add dependencies for conda
* Change to README
* Update README
* Use site data published in ks32
* Add running the configurations
* .gitignore files from install
* Fix: re-building source code
* Add building CABLE
* Fix setup with setup.cfg
* Moved all setup phases into a class
* Add setup for compilation
* Fix issue with configuration file format
* simplify the file structure
* Attempt to build for command line
* Define input config in yaml
* Create min. work directory tree
* Modify command line args
* Add simple command line
* Clean start
* Update .gitignore
* Add LICENSE
* separate inputs by type of input
* Update README.md
* Update README.md
* Forgotten loading the python module
* To run on multi-processors
* Some typos in the README file
* More docs updates to make things simpler for users
* Updated documentation
* Use hh5 environments in qsub script
* Load hh5 conda environment in qsub script

v0.1
----

* Serial runs only work for now
* Fix imports
* update all from optparse to argparse
* Black formatting
* A few missing things for mpi exe. name
* Add "-mpi" to exe name if compiled with mpi
* "Fix" error handling for subprocess. To update
* Import submodules properly
* Update to argparse. Remove append to path for imports
* To work with updated build script for CABLE
* Update how to find compilation options for Gadi
* Simplify check for NCI machine
* Update to Argparse from optparse
* Don't append to path for imports
* replace "is" with "=="
* fixes for jhan
* removed date
* fixed sys exit
* Added xarray
* fixed dates issue
* fixed for jhan
* fixed for share branch
* fixed option to pass user branch
* test more cores
* test more cores
* Changed call
* test
* test
* reduced core calls
* added unload openmpi
* changed logging options
* Added netcdf unload cmd
* fixed filename
* fixed seasonal plot wrapper
* Fixed logic for qsub
* fixed mpi
* fix for unused cores
* fix for unused cores
* changed default mpi
* removed clean
* fixed cmd line parser
* matched other run script
* matched other run script
* Added flag to build sep of get
* fixed literals
* removed qsub
* Changed tests
* Added NCI hint
* fixed literals issue
* fix for gadi modules
* fix for ifort
* fix for ifort
* fix for ifort
* fix for ifort
* fix for ifort
* fix for ifort
* fix for ifort
* added debug
* added debug
* fixed literal issue?
* test for gadi fi
* fixed storm server paths
* fixed storm server paths
* Added missing lib
* Added missing lib
* Added catch for anna
* fixed typo
* Changed storm paths
* clean up
* filename cleanup
* Added plumber path for storm
* changed trunk name
* blah
* added unplumbed spatial
* Added spatial file, needs editing
* fixed new name
* renamed file
* Added -s flag
* Added -s flag
* fixed met path
* fixed met path
* fixed met path
* fixed met path
* fixed to get it working again
* fixed paths
* fixed typos
* changed pbs options
* updated met files
* rolled back update utils
* updated utils
* updated text to include mention of integration
* fixed paths for gadi
* changed nc path
* changed paths for gadi
* changed paths for gadi
* changed paths for gadi
* Added note on integration branch
* Commented out old ver
* Added newline
* Changed subprocess call to one that automatically raises exception with error
* Tweaked multiprocessing code
* Tweaked multiprocessing code
* Fixed met flag
* Tweaked options
* Added sci config
* Added sci config
* Added further sci configs
* tweaked usage request
* Removed removal of src
* Removed file
* Fixed ignore
* Fixed ignore
* Changed path
* Fixed stupid logic
* Changed qsub options
* Added comment
* Turned on MPI
* Removed date from Trunk
* Made integration the default
* Made integration the default
* Added script to gen integration branch
* Added missing numpy
* Fixed home bug
* Added password SVN option
* Added missing dict check
* Added missing dict check
* Put back missings args
* Fixed options typo
* Added additional check
* Further edits
* Further edits
* Further edits
* Fixed readme
* removed wrapper script
* Split and reorganised files to work on PBS
* Renamed file
* Renamed file
* Removed correct command for non empty dir
* Removed debug stuff
* Added initial clean up of src dir
* Fixed netcdf load
* Fixed netcdf load
* Fixed netcdf load
* added debug statements
* added debug statements
* added debug statements
* added pbs wd
* added pbs wd
* added pbs wd
* added pbs wd
* various fixes to nci script
* changed raijin testing
* changed raijin testing
* Added qsub script
* Changed raijin met path
* Fixed raijin netcdf libs
* Changed paths on storm server to match old plumber files
* Changed paths for storm server
* Changed paths for storm server
* Changed paths for storm server
* Changed paths for storm server
* Added case for imac
* Added repo id and site id, for MPI stuff
* Fixed Matplotlib Deprecation Warning
* Moved paths bit to reduce immediate options
* Moved paths bit to reduce immediate options
* Added different NC paths
* Added different met paths
* Added different met paths
* Added functionality to not use the trunk
* Added options for rajin storm
* Added options for rajin storm

src/benchcab/config.py
# Copyright 2022 ACCESS-NRI and contributors. See the top-level COPYRIGHT file for details.
# SPDX-License-Identifier: Apache-2.0

"""A module containing all *_config() functions."""
import os
from pathlib import Path

import yaml
import copy
from cerberus import Validator
import base64
import hashlib
import benchcab.utils as bu
from benchcab.internal import MEORG_PROFILE
from benchcab import internal
from benchcab.utils.repo import create_repo
from benchcab.model import Model
from typing import Optional


class ConfigValidationError(Exception):
    """When config doesn't match with the defined schema."""

    def __init__(self, validator: Validator):
        """.

        Parameters
        ----------
        validator: cerberus.Validator
            A validation object that has been used and has the errors attribute.

        """
        # Nicely format the errors.
        errors = [f"{k} = {v}" for k, v in validator.errors.items()]

        # Assemble the error message and
        msg = "\n\nThe following errors were raised when validating the config file.\n"
        msg += "\n".join(errors) + "\n"

        # Raise to super.
        super().__init__(msg)


def validate_config(config: dict) -> bool:
    """Validate the configuration dictionary.

    Parameters
    ----------
    config : dict
        Dictionary of configuration loaded from the yaml file.

    Returns
    -------
    bool
        True if valid, exception raised otherwise.

    Raises
    ------
    ConfigValidationError
        Raised when the configuration file fails validation.

    """
    # Load the schema
    schema = bu.load_package_data("config-schema.yml")

    # Create a validator
    v = Validator(schema)

    # Validate
    is_valid = v.validate(config)

    # Valid
    if is_valid:
        return True

    # Invalid
    raise ConfigValidationError(v)


def read_optional_key(config: dict):
    """Fills all optional keys in config if not already defined.

    The default values for most optional keys are loaded from `internal.py`
    Note: We need to ensure that the `name` key for realisations exists for
    other modules, but it doesn't have a default value.  So we set it to
    `None` by default.

    Parameters
    ----------
    config : dict
        The configuration file with, or without optional keys

    """
    if "project" not in config:
        config["project"] = os.environ.get("PROJECT", None)

    if "realisations" in config:
        for r in config["realisations"]:
            r["name"] = r.get("name")

    config["science_configurations"] = config.get(
        "science_configurations", internal.DEFAULT_SCIENCE_CONFIGURATIONS
    )

    # Default values for spatial
    config["spatial"] = config.get("spatial", {})

    config["spatial"]["met_forcings"] = config["spatial"].get(
        "met_forcings", internal.SPATIAL_DEFAULT_MET_FORCINGS
    )

    config["spatial"]["payu"] = config["spatial"].get("payu", {})
    config["spatial"]["payu"]["config"] = config["spatial"]["payu"].get("config", {})
    config["spatial"]["payu"]["args"] = config["spatial"]["payu"].get("args")

    # Default values for fluxsite
    config["fluxsite"] = config.get("fluxsite", {})

    config["fluxsite"]["multiprocess"] = config["fluxsite"].get(
        "multiprocess", internal.FLUXSITE_DEFAULT_MULTIPROCESS
    )
    config["fluxsite"]["experiment"] = config["fluxsite"].get(
        "experiment", internal.FLUXSITE_DEFAULT_EXPERIMENT
    )
    config["fluxsite"]["pbs"] = internal.FLUXSITE_DEFAULT_PBS | config["fluxsite"].get(
        "pbs", {}
    )

    config["codecov"] = config.get("codecov", False)

    return config


def is_valid_meorg_output_name(name: str) -> Optional[str]:
    """Validate model output name against github issue standards.

    Standard: <digit>-<words-sep-by-dashes>

    Parameters
    ----------
    name: str
        The model output name

    Returns
    -------
    Optional[str]
        If model output name does not meet standard, then return error message

    """
    if len(name) == 0:
        return "Model output name is empty\n"

    msg = ""

    if len(name) > 50:
        msg += "The length of model output name must be shorter than 50 characters. E.g.: 1-length-is-20-chars\n"

    if " " in name:
        msg += "Model output name cannot have spaces. It should use dashes (-) to separate words. E.g. 123-word1-word2\n"

    name_keywords = name.split("-")

    if not name_keywords[0].isdigit():
        msg += "Model output name does not start with number, E.g. 123-number-before-word\n"

    if len(name_keywords) == 1:
        msg += "Model output name does not contain keyword after number, E.g. 123-keyword\n"

    if msg == "":
        return None

    return f"Errors present when validating model output name:\n{msg}"


def add_meorg_output_name(config: dict):
    """Determine model output name from realisations.

    Parameters
    ----------
    config : dict
        The configuration file with optional keys

    """
    # pure function
    config = copy.deepcopy(config)

    mo_names = [True for r in config["realisations"] if r.get("meorg_output_name")]

    if len(mo_names) > 1:
        msg = "More than 1 value set as true"
        raise AssertionError(msg)

    mo_names = ""
    for r in config["realisations"]:
        # `meorg_output_name` decided either via `name` parameter in a realisation,
        # otherwise via `Repo` branch name
        repo = create_repo(
            spec=r["repo"],
            path=internal.SRC_DIR / (r["name"] if r.get("name") else Path()),
        )
        mo_name = Model(repo, name=r.get("name")).name

        mo_names += mo_name
        if r.pop("meorg_output_name", None):
            msg = is_valid_meorg_output_name(mo_name)
            if msg is not None:
                raise Exception(msg)

            config["meorg_output_name"] = mo_name

    if "meorg_output_name" in config:
        user = os.getenv("USER")
        mo_name_hash_input = f"{mo_names}{MEORG_PROFILE['id']}{user}"
        # hash in bytes form
        mo_name_hash_b = hashlib.sha1(mo_name_hash_input.encode())
        # Convert to str and take first 6 characters
        mo_name_hash = base64.b32encode(mo_name_hash_b.digest()).decode()[:6]
        config["meorg_output_name"] += f"_{mo_name_hash}"

    return config


def read_config_file(config_path: str) -> dict:
    """Load the config file in a dict.

    Parameters
    ----------
    config_path : str
        Path to the configuration file.

    Returns
    -------
    dict
        Configuration dict

    """
    # Load the configuration file.
    with Path.open(Path(config_path), "r", encoding="utf-8") as file:
        config = yaml.safe_load(file)

    return config


def read_config(config_path: str) -> dict:
    """Reads the config file and returns a dictionary containing the configurations.

    Parameters
    ----------
    config_path : str
        Path to the configuration file.
    is_meorg: str
        Whether workflow includes meorg job submission. If true, determine the model output name

    Returns
    -------
    dict
        Validated configuration dict, with default optional parameters if not specified in file.

    Raises
    ------
    ConfigValidationError
        Raised when the configuration file fails validation.

    """
    # Read configuration file
    config = read_config_file(config_path)
    # Populate configuration dict with optional keys
    config = read_optional_key(config)
    # Validate.
    validate_config(config)
    config = add_meorg_output_name(config)
    return config
